pkg/unidling/api/types.go

```diff
@@ -0,0 +1,20 @@
+package api
+
+import (
+	kextapi "k8s.io/kubernetes/pkg/apis/extensions"
+)
+
+const (
+	IdledAtAnnotation      = "idling.alpha.openshift.io/idled-at"
+	UnidleTargetAnnotation = "idling.alpha.openshift.io/unidle-targets"
+	NeedPodsReason         = "NeedPods"
+)
+
+// RecordedScaleReference is a SubresourceReference to a scale subresource that also
+// has the previous replica count recorded
+type RecordedScaleReference struct {
+	// TODO: update this to a CrossVersionObjectReference once we pull in a Kube version
+	// with an unversioned apis/autoscaling
+	kextapi.SubresourceReference `json:",inline"`
```

Comment:
> Feel free to correct me if I'm wrong here, but that suggestion seems to completely go against existing API design -- all the other reference types, AFAICT, refer to Kind instead of Resource and APIVersion in the slash-separated form (e.g. ObjectReference, OwnerReference, and group-versions are written as APIVersion in TypeMeta as well).

Resource and kinds are different. Scale is a great example.  Scale is a kind, used by many resources.  We also have Binding which could logically be re-used.  If we accept that polymorphic endpoints take Kinds (they have to or the system doesn't work), then kind does not always map one to one to resource.  

Note that this system also failed for the origin and kube rbac types, where we wanted one Kind for multiple resources and people's bad assumptions on those one to one mappings (which don't even hold today) broke us.

`APIVersion` is used a slash delimited field, but in the case of a reference are you likely to care which version is used?  Almost never.  If both servers speak v2, use v2.  If they both speak v1, use it.  The only case where a distinction matters is when a resource changes meaning between versions.  This results in an APIVersion field that should almost never contain a version.  That's a pretty bad misnomer.



> Furthermore, the scale client interfaces refer to Kind as well, so if we took a resource, we'd just have to convert it back to kind anyway (furthermore, I don't think we want to explicitly state that we're referring to the scale subresource here -- just an object that happens to have a scale subresource. I'd much rather use

The current scale implementation is broken.  It takes a kind and assumes that a lowercase and pluralization will give it a resource.  Obviously that fails in the general case since the very first thing we did between versions was change Minion to Node so a Minion kind has a nodes resource.  There is a way to handle the mapping, but again, its a one to many, so you can end up failing.  You can see comments in the pulls establishing the current implementation point out these flaws.

tldr Kube did it wrong and has produced code that doesn't work right for the types they have in existence today.